### PR TITLE
Big update, replace all linked-list by dynamic array

### PR DESCRIPTION
**Pull request by luohaha**
Use one array to replace four list. Each page in array has five status:
 ==================
 @PAGE_SLEEP : this page isn't used now.
 @PAGE_DOWNLOAD_WAIT : this page's task url is assigned now.
 @PAGE_DOWNLOAD_RUNNING : this page's goal html is downloading.
 @PAGE_PROCESS_WAIT : this page's goal html is waited to process.
 @PAGE_PROCESS_RUNNING : this page's goal html is processing now.

find more details in cs_page_queue.h

### CODE REVIEW

**Comment by luohaha** [comment]
This is a big update. It change many files, so ask me any thing that you don't understand.

File: core/downloader/downloader.c [hunk 1]
```diff
@@ -12,22 +12,24 @@
   @ss : input pointer
 **/
 size_t save_data(void *ptr, size_t size, size_t nmemb, void *ss) {
-  cs_task_t *save = (cs_task_t*)ss;
-  size_t count = save->data->count;
+  //cs_task_t *save = (cs_task_t*)ss;
+  cs_page *page = (cs_page*)ss;
+  //size_t count = save->data->count;
```
Comment (AntonioCS) [inline core/downloader/downloader.c]: You should remove the commented code. If its not needed its not doing anything.

File: core/downloader/downloader.c [hunk 2]
```diff
@@ -12,22 +12,24 @@
   @ss : input pointer
 **/
 size_t save_data(void *ptr, size_t size, size_t nmemb, void *ss) {
```
Comment (AntonioCS) [inline core/downloader/downloader.c]: If you are doing a major rework try adding the cspider_ namespace to the functions.

File: core/includes/CS.h [hunk 3]
```diff
@@ -27,6 +27,7 @@
 the max number of the buffers
 */
 #define BUFFER_MAX_NUMBER 1024
+#define bool char
```
Comment (AntonioCS) [inline core/includes/CS.h]: Can't you just use <stdbool.h> for all boolean related things?

File: core/pageProcesser/user_api.c [hunk 4]
```diff
@@ -60,9 +60,14 @@ void addUrl(cspider_t *cspider, char *url) {
     PANIC(reUrl);
     
     strncpy(reUrl, url, len);
-    uv_rwlock_wrlock(cspider->lock);
-    createTask(cspider->task_queue, reUrl);
-    uv_rwlock_wrunlock(cspider->lock);
+    //createTask(cspider->task_queue, reUrl);
+    cs_page *page = get_page(cspider->page_queue);
+    set_url(page, reUrl);
+    /**
```
Comment (AntonioCS) [inline core/pageProcesser/user_api.c]: You sometimes put one line comments with // and sometimes use /* */. Please be consistent :)

File: sample/douban.c [hunk 5]
```diff
@@ -0,0 +1,57 @@
+#include<cspider/spider.h>
+typedef struct {
+  char *getTitle[20];
+  char *getDesc[20];
+  int size;
+} Movie;
+//开始入口的 url
```
Comment (AntonioCS) [inline sample/douban.c]: No chinese please!!!

File: sample/douban.c [hunk 6]
```diff
@@ -0,0 +1,57 @@
+#include<cspider/spider.h>
+typedef struct {
```
Comment (AntonioCS) [inline sample/douban.c]: As this is an example code, maybe not compress the code. Use \n wisely

File: sample/douban.c [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+void s(void *str, void *user_data) {
+  Movie *get = (Movie*)str;
+  FILE *file = (FILE*)user_data;
+  int i;
+  for (i = 0; i < get->size; i++) {
+    fprintf(file, "名称:%s\n", get->getTitle[i]);
+    fprintf(file, "类别:%s\n", get->getDesc[i]);
+  }
+}
+int main() {
```
Comment (AntonioCS) [inline sample/douban.c]: Maybe also add a few more comments on important lines so that it is clear what is happening.

**Comment by luohaha** [comment]
@AntonioCS Thanks!

**Comment by mzer0-yu** [comment]
@luohaha 
bool is 2 bytes sometimes, because the C standard defines the behavior of bool type conversions.
1) bool to other type.
e.g. any_type foo; 
       foo = true; // foo == 1;
       foo = false; // foo == 0;
2) other type to bool.
e.g. bool foo;
       foo = x; // if x == 0 then foo = false; else foo = true;
------------------------------------------------------------------------------------------------------------------------------------
So it doesn't matter what the length of bool is. Even if it is 4 bytes, or 8 bytes, because the conversion is independent from its length. Changing the length of bool is aim to memory alignment. In Windows(MSVC), it may define BOOL to CHAR because of the ABI(Application Binary Interface).

**Pull request merged**